packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/device/DeviceHandler.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.amplify.amplify_auth_cognito.device
+
+import com.amazonaws.amplify.amplify_auth_cognito.AuthErrorHandler
+import com.amazonaws.amplify.amplify_auth_cognito.base.AtomicResult
+import com.amazonaws.mobile.client.AWSMobileClient
+import com.amazonaws.mobile.client.Callback
+import com.amazonaws.mobile.client.results.ListDevicesResult
+import com.amplifyframework.auth.AuthDevice
+import com.amplifyframework.auth.AuthException
+import com.amplifyframework.auth.cognito.util.CognitoAuthExceptionConverter
+import com.amplifyframework.core.Amplify
+import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
+import kotlinx.coroutines.*
+
+/**
+ * Handles method calls for the Devices API.
+ */
+class DeviceHandler(private val errorHandler: AuthErrorHandler) :
+    MethodChannel.MethodCallHandler {
+    companion object {
+        /**
+         * Methods this handler supports.
+         */
+        private val methods = listOf("rememberDevice", "forgetDevice", "fetchDevices")
+
+        /**
+         * Whether this class can handle [method].
+         */
+        fun canHandle(method: String): Boolean = methods.contains(method)
+    }
+
+    /**
+     * Scope for running asynchronous tasks.
+     */
+    private val scope = CoroutineScope(Dispatchers.IO) + CoroutineName("DeviceHandler")
+
+    @Suppress("UNCHECKED_CAST")
+    override fun onMethodCall(call: MethodCall, _result: MethodChannel.Result) {
+        val result = AtomicResult(_result)
+        when (call.method) {
+            "fetchDevices" -> fetchDevices(result)
+            "rememberDevice" -> rememberDevice(result)
+            "forgetDevice" -> {
+                val deviceJson = (call.arguments as? Map<*, *> ?: emptyMap<String, Any?>()) as Map<String, Any?>
```

Comment:
It's kind of arbitrary, I suppose. I prefer the `forgetDevice` method to take in an AuthDevice and not a Map or an Any.

Reply:
If `forgetDevice` wasn't already coupled to the method channel invocation, I would agree that passing an AuthDevice makes far more sense, but since `forgetDevice` takes an argument of `MethodChannel.Result`, it makes sense to me to pass in all the arguments from that method channel call (call.arguments). It is a `nit` comment though so this definitely isn't blocking.

Reply:
Ah, I see your point, now. Thanks for clarifying. I think this is a matter of preference - I can see the argument for both options now, though. I do still prefer it this way and the reason is that I tend to think of result as being an interface which can be mocked or extended whereas the device argument is an ADT and by replacing AuthDevice with Map or Any, we lose the type hierarchy and any type safety in the method signature.

Again, seems like a personal preference but just wanted to expand too on my rationale.

Reply:
Ideally, we would have multiple layers to this architecture - one dedicated to decoding and one dedicated to working with decoded types. I think that's what is creating the disparity for us both, since we are dancing in the middle here.